mono/metadata/cominterop.c

```diff
@@ -3699,32 +3669,46 @@ mono_cominterop_get_com_interface (MonoObject *object, MonoClass *ic, MonoError
 			return NULL;
 		klass = mono_object_class (object);
 		if (!mono_class_is_transparent_proxy (klass)) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "Class is not transparent");
 			return NULL;
 		}
 
 		real_proxy = ((MonoTransparentProxy*)object)->rp;
 		if (!real_proxy) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "RealProxy is null");
 			return NULL;
 		}
 
 		klass = mono_object_class (real_proxy);
 		if (klass != mono_class_get_interop_proxy_class ()) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "Object is not a proxy");
 			return NULL;
 		}
 
 		if (!((MonoComInteropProxy*)real_proxy)->com_object) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "Proxy points to null COM object");
 			return NULL;
 		}
 
+		if (icall)
+			return ((MonoComInteropProxy*)real_proxy)->com_object->iunknown;
 		void* com_itf = cominterop_get_interface_checked (((MonoComInteropProxy*)real_proxy)->com_object, ic, error);
 		return com_itf;
 	}
 	else {
+		if (icall)
+			ic = mono_class_get_iunknown_class ();
 		void* ccw_entry = cominterop_get_ccw_checked (object, ic, error);
+		if (icall)
+			mono_error_set_pending_exception (error);
```

Comment:
I think it's better to have the call to `mono_error_set_pending_exception` in`ves_icall_System_Runtime_InteropServices_Marshal_GetIUnknownForObjectInternal` not in `..._internal`. It will make the icall more like others and will eventually make it easier to convert it to handles.

Reply:
I think it should be asis to perfectly preserve the old meaning. Coop conversion will remove these 2 lines either way.

Reply:
mono_error_set_pending_exception could be moved to caller since it is last, and the other branch of the arm cannot cause it to do anything. Though asis is slightly more obviously semantic preserving, and these two lines should go away soon either way.

The other not_reached vs. error cannot be so obviously or "smally" moved as I understand.
For example, we could return an out "invalid_operation string" for caller to not_reached or mono_error_set_invalid_operation. Another characteristic of the current one is an easier to read diff where at least one path has no change, just conditional adds that won't occur.